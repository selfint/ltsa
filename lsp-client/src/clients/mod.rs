mod stdio;

pub use stdio::{child_client, stdio_client};
