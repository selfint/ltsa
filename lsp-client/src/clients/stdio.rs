use tokio::{
    io::{AsyncBufReadExt, AsyncWriteExt, BufReader},
    process::{ChildStderr, ChildStdin, ChildStdout},
    sync::mpsc::{unbounded_channel, UnboundedSender},
    task::JoinHandle,
};

use crate::client::Client;

pub fn stdio_client(
    mut stdin: ChildStdin,
    stdout: ChildStdout,
    stderr: ChildStderr,
) -> (Client, Vec<JoinHandle<()>>) {
    let (client_tx, mut client_rx) = unbounded_channel::<String>();
    let (server_tx, server_rx) = unbounded_channel();

    let server_input_handle = tokio::spawn(async move {
        while let Some(msg) = client_rx.recv().await {
            stdin.write_all(msg.as_bytes()).await.unwrap();
        }
    });

    let server_output_handle = stdout_proxy(BufReader::new(stdout), server_tx);

    let mut stderr_lines = BufReader::new(stderr).lines();
    let server_error_handle = tokio::spawn(async move {
        while let Ok(Some(line)) = stderr_lines.next_line().await {
            eprintln!("Got err from server: {}", line);
        }
    });

    let client = Client::new(client_tx, server_rx);

    (
        client,
        vec![
            server_input_handle,
            server_output_handle,
            server_error_handle,
        ],
    )
}

fn stdout_proxy(mut rx: BufReader<ChildStdout>, tx: UnboundedSender<String>) -> JoinHandle<()> {
    tokio::spawn(async move {
        let mut next_content_length = None;
        let mut next_content_type = None;

        loop {
            let mut line = String::new();
            rx.read_line(&mut line).await.unwrap();

            let words = line.split_ascii_whitespace().collect::<Vec<_>>();
            match (
                words.as_slice(),
                &mut next_content_length,
                &mut next_content_type,
            ) {
                (["Content-Length:", content_length], None, None) => {
                    next_content_length = Some(content_length.parse().unwrap())
                }
                (["Content-Type:", content_type], Some(_), None) => {
                    next_content_type = Some(content_type.to_string())
                }
                (["Content-Type:", content_type, "charset=utf8"], Some(_), None) => {
                    next_content_type = Some(content_type.to_string())
                }
                ([], Some(content_length), _) => {
                    let mut content = Vec::with_capacity(*content_length);
                    let mut bytes_left = *content_length;
                    while bytes_left > 0 {
                        let read_bytes = rx.read_until(b'}', &mut content).await.unwrap();
                        bytes_left -= read_bytes;
                    }

                    let content = String::from_utf8(content).unwrap();
                    tx.send(content).unwrap();

                    next_content_length = None;
                    next_content_type = None;
                }
                other => panic!("Got unexpected stdout: {:?}", other),
            };
        }
    })
}
