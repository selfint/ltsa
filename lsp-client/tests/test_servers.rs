use lsp_client::clients;
use lsp_types::{
    notification::Initialized, request::Initialize, InitializeParams, InitializedParams,
};
use std::process::Stdio;
use tokio::process::{Child, Command};

fn start_rust_analyzer() -> Child {
    Command::new("rust-analyzer")
        .stdin(Stdio::piped())
        .stdout(Stdio::piped())
        .stderr(Stdio::piped())
        .spawn()
        .expect("failed to start rust analyzer")
}

fn start_solidity_ls() -> Child {
    Command::new("solidity-ls")
        .arg("--stdio")
        .stdin(Stdio::piped())
        .stdout(Stdio::piped())
        .stderr(Stdio::piped())
        .spawn()
        .expect("failed to start rust analyzer")
}

fn start_python_language_server() -> Child {
    Command::new("pyls")
        .stdin(Stdio::piped())
        .stdout(Stdio::piped())
        .stderr(Stdio::piped())
        .spawn()
        .expect("failed to start rust analyzer")
}

macro_rules! test_server {
    ($start_server_fn:ident) => {
        let mut child = $start_server_fn();

        let stdin = child.stdin.take().unwrap();
        let stdout = child.stdout.take().unwrap();
        let stderr = child.stderr.take().unwrap();

        let (client, handles) = clients::stdio_client(stdin, stdout, stderr);

        let init_resp = client
            .request::<Initialize>(InitializeParams::default())
            .await;

        insta::assert_debug_snapshot!(init_resp);

        client.notify::<Initialized>(InitializedParams {}).unwrap();

        for handle in handles {
            handle.abort();
        }
    };
}

#[tokio::test]
async fn test_rust_analyzer() {
    test_server!(start_rust_analyzer);
}

#[tokio::test]
async fn test_solidity_ls() {
    test_server!(start_solidity_ls);
}

#[tokio::test]
async fn test_python_language_server() {
    test_server!(start_python_language_server);
}
