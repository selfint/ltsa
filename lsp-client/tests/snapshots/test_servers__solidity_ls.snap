---
source: lsp-client/tests/test_servers.rs
expression: init_resp
---
Ok(
    Ok(
        InitializeResult {
            capabilities: ServerCapabilities {
                position_encoding: None,
                text_document_sync: Some(
                    Kind(
                        Incremental,
                    ),
                ),
                selection_range_provider: None,
                hover_provider: Some(
                    Simple(
                        true,
                    ),
                ),
                completion_provider: Some(
                    CompletionOptions {
                        resolve_provider: None,
                        trigger_characters: Some(
                            [
                                ".",
                            ],
                        ),
                        all_commit_characters: None,
                        work_done_progress_options: WorkDoneProgressOptions {
                            work_done_progress: None,
                        },
                        completion_item: None,
                    },
                ),
                signature_help_provider: Some(
                    SignatureHelpOptions {
                        trigger_characters: Some(
                            [
                                "(",
                            ],
                        ),
                        retrigger_characters: None,
                        work_done_progress_options: WorkDoneProgressOptions {
                            work_done_progress: None,
                        },
                    },
                ),
                definition_provider: Some(
                    Left(
                        true,
                    ),
                ),
                type_definition_provider: None,
                implementation_provider: None,
                references_provider: Some(
                    Left(
                        true,
                    ),
                ),
                document_highlight_provider: None,
                document_symbol_provider: None,
                workspace_symbol_provider: None,
                code_action_provider: None,
                code_lens_provider: None,
                document_formatting_provider: Some(
                    Left(
                        true,
                    ),
                ),
                document_range_formatting_provider: None,
                document_on_type_formatting_provider: None,
                rename_provider: Some(
                    Left(
                        true,
                    ),
                ),
                document_link_provider: None,
                color_provider: None,
                folding_range_provider: None,
                declaration_provider: None,
                execute_command_provider: None,
                workspace: None,
                call_hierarchy_provider: None,
                semantic_tokens_provider: None,
                moniker_provider: None,
                inline_value_provider: None,
                inlay_hint_provider: None,
                linked_editing_range_provider: None,
                experimental: None,
            },
            server_info: None,
        },
    ),
)
